dialogue-target/src/main/java/com/palantir/dialogue/Endpoint.java

```diff
@@ -25,7 +28,23 @@
  * as well as the types of the request and response body.
  */
 public interface Endpoint {
-    void renderPath(Map<String, String> params, UrlBuilder url);
+    /**
+     * {@link #renderPath(ListMultimap, UrlBuilder)} is preferred, however at least one {@code renderPath} must be
+     * implemented to avoid infinite recursion.
+     */
+    default void renderPath(Map<String, String> params, UrlBuilder url) {
+        renderPath(
+                params.isEmpty()
+                        ? ImmutableListMultimap.of()
+                        : ImmutableListMultimap.<String, String>builder()
+                                .putAll(params.entrySet())
+                                .build(),
+                url);
+    }
+
+    default void renderPath(ListMultimap<String, String> params, UrlBuilder url) {
```

Comment:
Tell me if I'm wrong: but maybe instead of implementing this method, we should instead have:

```
PathTemplate pathTemplate();
```

If we had it this way, it would not have been a break at all?

Reply:
As an alternative, we could have Endpoint2?

Reply:
I think one would require the other, I’m not sure we could generate the template for backcompat on existing endpoints.

I don’t recall why we provide a method rather than an accessor for the template though, exposing the template would be cleaner. Rolling out Endpoint2 would cause a lot of churn given how widely used Endpoint is.

Reply:
It's generate code though, does it matter?

Reply:
Endpoint2 would require overloads everywhere in this project for backcompat, right?

I just rolled out of bed and might not be groking something.

Reply:
Ah yes, I see. I thought it would only be relevant to the url generation and maybe some metrics, but it's used all over. I guess we could major rev, but probably not worth the pain.